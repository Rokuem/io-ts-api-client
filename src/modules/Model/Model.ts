import * as t from 'io-ts';
import reporter from 'io-ts-reporters';
import { Either, isRight } from 'fp-ts/lib/Either';
import { TypedEmitter } from '../TypedEmitter/TypedEmmiter';
import { addedDiff } from 'deep-object-diff';
import { ValidationOptions } from '../types';
import { deepMerge } from '../../helpers/deepMerge';

/**
 * Simple io-ts model with a validator. This is used to safeguard against API changes and to provide a easy solution for type checking it.
 */
export class Model<Schema extends t.Any = any, SampleKey extends string = never, Samples extends t.TypeOf<Schema> = t.TypeOf<any>> {
  /**
   * Emitter for modal events.
   */
  public static emitter = new TypedEmitter<{
    'before-validation': (model: string) => void;
    'validation-success': (model: string) => void;
    'validation-error': (model: string, error: Error) => void;
    'after-validation': (model: string) => void;
    'extra-keys-detected': (model: string, error: Error) => void;
  }>();

  /**
   * a name to call the model during logs and errors.
   */
  public name: string;

  /**
   * @deprecated - use model.schema instead.
   */
  public base: Schema;

   /**
   * IO-TS interface used for the model.
   *
   * you can use this to extend other interfaces.
   */
  public schema!: Schema;

  /**
   * Ts interface of the model.
   * 
   * !Warn: Do not use this as an value.
   * @since 0.7.4
   */
  public tsInterface!: t.TypeOf<Schema>

  public samples!: Record<SampleKey, Samples>;

  public constructor(config: {
    /**
     * The name of the model.
     */
    name: string;
    /**
     * The io-ts model.
     */
    schema: Schema;
    samples?: Record<SampleKey, Samples>;
  }) {
    this.name = config.name;
    this.base = config.schema;
    this.schema = config.schema;
    this.samples = config.samples || {} as any;
  }

  /**
   * If not an empty string, this model belongs to an operation.
   */
  public operation = '';

  /**
   * Checks if there are any extra keys in the response received.
   */
  private checkForExtraKeys(options: {
    result: Either<t.Errors, any>;
    target: any;
    debug: boolean;
    throwErrors: boolean;
  }) {
    const { result, target } = options;

    if (!('right' in result) && options.debug) {
      console.warn('Extra fields check skipped due to failed validation');
      return;
    }

    if ('props' in this.base && 'right' in result) {
      const filteredDecode = t.exact(this.base as any).decode(target);
      const resultJson = JSON.stringify(result.right);

      if (!('right' in filteredDecode)) {
        throw new Error('Something went wrong while checking for extra keys');
      }

      const filtered = filteredDecode.right;
      const filteredJson = JSON.stringify(filtered);

      if (filteredJson !== resultJson) {
        const diff = addedDiff(filtered, result.right);
        const errorMessage = `${
          this.operation ? `[${this.operation}] ` : ''
        }Detected extra properties in model "${this.name}": ${JSON.stringify(
          diff,
          null,
          '  '
        )}`;

        Model.emitter.emit('extra-keys-detected', this.name, new Error(errorMessage));

        if (options.throwErrors) {
          throw new Error(errorMessage);
        } else if (options.debug) {
          console.error(errorMessage);
        }
      }

      return filtered;
    }
  }

  private handleInvalidResult(options: {
    result: Either<t.Errors, any>;
    target: any;
    throwErrors: boolean;
    debug: boolean;
  }) {
    const validationErrors = reporter.report(options.result);

    for (const error of validationErrors) {
      const validationMessage = `ERROR - ${
        this.operation ? `[${this.operation}]` : ''
      }[${this.name}] VALIDATION FAILED: ${error} at: \n ${JSON.stringify(
        options.target,
        null,
        2
      )}`;

      Model.emitter.emit('validation-error', this.name, new Error(validationMessage));

      if (options.debug) {
        console.error(validationMessage);
      }

      if (options.throwErrors) {
        throw new Error(validationMessage);
      }
    }
  }

  private handleValidResult(
    target: any,
    options: ValidationOptions
  ) {
    if (options.debug) {
      console.log(
        `SUCCESS - ${this.operation ? `[${this.operation}]` : ''}[${
          this.name
        }]: Target is valid!`,
        {
          target,
        }
      );
    }

    Model.emitter.emit('validation-success', this.name);
  }

  /**
   * Validates the model. Throws error if failed in strict mode.
   */
  public validate(
    target: any,
    options: ValidationOptions
  ) {
    Model.emitter.emit('before-validation', this.name);

    let result = this.base.decode(target);
    const isValid = isRight(result);

    if (options.strictTypes) {
      const filtered = this.checkForExtraKeys({
        result,
        target,
        debug: options.debug,
        throwErrors: options.throwErrors,
      });

      if (filtered) {
        result = {
          ...result,
          right: filtered,
        } as any;
      }
    }

    if (!isValid) {
      try {
        this.handleInvalidResult({
          target,
          result,
          debug: !!options.debug,
          throwErrors: !!options.throwErrors,
        });
      } catch (e) {
        Model.emitter.emit('after-validation', this.name);
        throw e;
      }
    } else {
      this.handleValidResult(target, options);
    }

    Model.emitter.emit('after-validation', this.name);

    return 'right' in result ? result.right : target;
  }

  /**
   * Asserts that target is valid for this model.
   */
  public assert(target: any, options?: { 
    /**
     * Fails the assertion in case the target has extra properties.
     */
    strict?: boolean;
    /**
     * Debug the validation process.
     */
    debug?: boolean;
  }): target is ModelInterface<this> {
    try {
      this.validate(target, {
        debug: !!options?.debug,
        throwErrors: true,
        strictTypes: !!options?.strict,
      });
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Creates an object that matches the model.
   * Useful for creating mocks.
   */
  public createSample<T extends t.TypeOf<Schema>>(sample: T) {
    return sample;
  }

  /**
   * Extend one of the declared samples in the model declaration.
   * Useful for tests.
   */
  public extendSample<T extends Partial<t.TypeOf<Schema>>, K extends SampleKey>(base: K, sample: T) {
    return deepMerge(this.samples[base], sample);
  }
}

export type ModelInterface<T extends Model<any>> = t.TypeOf<T['base']>;
