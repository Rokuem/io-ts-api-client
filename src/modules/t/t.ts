import * as types from 'io-ts';
import * as extraTypes from 'io-ts-types';

export const t = {
  ...types,
  ...extraTypes,
  /**
   * Creates null intersection with provided type.
   */
  nullable: <T extends types.Any>(type: T) => types.union([t.null, type]),
  /**
   * Creates intersection between required and optional props.
   * @since 0.7.4
   */
  schema: <T extends types.Props, P extends types.Props>(type: { required: T, optional: P }) => types.intersection([
    types.interface(type.required),
    types.partial(type.optional)
  ]),
  /**
   * Type for valid date strings.
   * @since 0.8.5
   */
  dateString: new types.Type<string, string, unknown>(
    'date string',
    (input: unknown): input is string => typeof input === 'string',
    (input, context) => {
      if (typeof input !== 'string') {
        return types.failure(input, context);
      }

      const date = new Date(input);

      if (isNaN(date.getTime())) {
        return types.failure(input, context);
      }

      return types.success(input);
    },
    types.identity
  ),
  /**
   * Type for valid urls.
   * @since 0.8.5
   */
  url: new types.Type<string, string, unknown>(
    'url',
    (input: unknown): input is string => typeof input === 'string',
    (input, context) => {
      if (typeof input !== 'string') {
        return types.failure(input, context);
      }
  
      try {
        // URL constructors fails with invalid urls.
        new URL(input);
        return types.success(input);
      } catch (e) {
        return types.failure(input, context);
      }
    },
    types.identity
  )
};