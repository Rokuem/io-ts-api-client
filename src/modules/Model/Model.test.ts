import Model from './Model';
import { t } from '../t/t';

describe('A model', () => {
  beforeEach(() => {
    Model.emitter.resetListeners();
    Model.throwErrors = false;
    Model.debug = false;
    jest.resetAllMocks();
  });

  describe('When created', () => {
    test('Should match the snapshot', () => {
      const model = new Model({
        name: 'snapshot model',
        schema: t.interface({
          key: t.literal('value'),
        }),
      });

      expect(model).toMatchSnapshot('Basic Model');
    });
  });

  describe('When validating', () => {
    test('Should emit success when the types match', () => {
      const testModelName = 'test model';
      const cb = jest.fn();
      const model = new Model({
        name: testModelName,
        schema: t.literal('value'),
      });

      Model.emitter.on('validation-success', (name) => {
        expect(name).toBe(testModelName);
        cb();
      });

      model.validate('value');

      expect(cb).toHaveBeenCalledTimes(1);
    });

    test('Should log success when the types match in debug mode', () => {
      let loggedMessage: any[] = [];
      console.log = jest.fn((...message: any[]) => {
        loggedMessage = message;
      });

      Model.debug = true;

      const modelName = 'success message test model';

      const model = new Model({
        name: modelName,
        schema: t.interface({
          foo: t.literal('bar'),
        }),
      });

      const target = {
        foo: 'bar',
      };

      model.validate(target);

      expect(console.log).toHaveBeenCalledTimes(1);
      expect(loggedMessage).toEqual(
        expect.arrayContaining([
          expect.stringContaining(modelName.toUpperCase()),
        ])
      );

      expect(loggedMessage).toMatchSnapshot('Validation Success Message');
    });

    test('Should log error when the types do not match in debug mode', () => {
      let loggedMessage: any[] = [];
      console.error = jest.fn((...message: any[]) => {
        loggedMessage = message;
      });

      Model.debug = true;

      const modelName = 'error message test model';

      const model = new Model({
        name: modelName,
        schema: t.interface({
          foo: t.literal('bar'),
        }),
      });

      model.validate({
        foo: 'bob',
      });

      expect(console.error).toHaveBeenCalledTimes(1);
      expect(loggedMessage).toMatchSnapshot('Validation Error Message');
      expect(loggedMessage).toEqual(
        expect.arrayContaining([
          expect.stringContaining(modelName.toUpperCase()),
        ])
      );
    });

    test('Should emit error when the types do not match', () => {
      const testModelName = 'test model';
      const cb = jest.fn();
      const model = new Model({
        name: testModelName,
        schema: t.interface({
          foo: t.literal('bar'),
        }),
      });
      Model.emitter.on('validation-error', (name, error) => {
        expect(name).toBe(testModelName);
        expect(error).toMatchSnapshot('Error message');
        cb();
      });

      model.validate({
        foo: 'bob',
      });

      expect(cb).toHaveBeenCalledTimes(1);
    });

    test('Should throw error when the types do not match in strict mode', () => {
      Model.throwErrors = true;
      const modelName = 'throw error test model';
      const key = 'foo';
      const wrongValue = 'bob';
      const correctValue = 'bar';

      const model = new Model({
        name: modelName,
        schema: t.interface({
          [key]: t.literal(correctValue),
        }),
      });

      expect(() => {
        try {
          model.validate({
            [key]: wrongValue,
          });
        } catch (error) {
          expect(error.message.includes(modelName)).toBe(true);
          expect(error.message.includes(wrongValue)).toBe(true);
          expect(error.message.includes(correctValue)).toBe(true);
          expect(error.message.includes(key)).toBe(true);
        }
      }).toThrow();
    });
  });
});
