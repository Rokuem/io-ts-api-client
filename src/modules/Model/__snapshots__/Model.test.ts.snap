// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A model When created Should match the snapshot: Basic Model 1`] = `
Model {
  "base": InterfaceType {
    "_tag": "InterfaceType",
    "decode": [Function],
    "encode": [Function],
    "is": [Function],
    "name": "{ key: \\"value\\" }",
    "props": Object {
      "key": LiteralType {
        "_tag": "LiteralType",
        "decode": [Function],
        "encode": [Function],
        "is": [Function],
        "name": "\\"value\\"",
        "validate": [Function],
        "value": "value",
      },
    },
    "validate": [Function],
  },
  "name": "snapshot model",
}
`;

exports[`A model When validating Should emit error when the types do not match: Error message 1`] = `"Expecting \\"bar\\" at foo but instead got: \\"bob\\""`;

exports[`A model When validating Should log error when the types do not match in debug mode: Validation Error Message 1`] = `
Array [
  "ERROR - [ERROR MESSAGE TEST MODEL] VALIDATION FAILED: Expecting \\"bar\\" at foo but instead got: \\"bob\\" at: 
 {
  \\"foo\\": \\"bob\\"
}",
]
`;

exports[`A model When validating Should log success when the types match in debug mode: Validation Success Message 1`] = `
Array [
  "SUCCESS - [SUCCESS MESSAGE TEST MODEL]: Target is valid!",
  Object {
    "target": Object {
      "foo": "bar",
    },
  },
]
`;
